stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -47,38 +42,33 @@ inline matrix_gpu cholesky_decompose_recursion(matrix_gpu& A,
  * original matrix \f$A\f$ is given by
  * <p>\f$A = L \times L^T\f$.
  * The Cholesky decomposition is computed on the GPU. This algorithm is
- * recursive, where The parameters <code>block</code>, <code>divisor</code>, and
- *  <code>min_block</code> act as tuning parameters for the recursive step of
- *  the GPU based Cholesky decompostion. The matrix is subset by the
- *  <code>block</code> size, and if the <code>block</code> size is less than
- * <code>min_block</code> then the cholesky decomposition on the GPU is computed
+ * recursive. The matrix is subset into a matrix of size
+ *  <code>A.rows() / 2</code>, and if the <code>block</code> size is less than
+ * 100 then the cholesky decomposition on the GPU is computed
  * using that submatrix. If <code>block</code> is greater than
- * <code>block_size</code> then <code>cholesky_decompose</code> is run again
- * with <code>block</code> equal to <code>block/divisor</code>. Once the
+ * 100 or <code>min_block</code> then <code>cholesky_decompose</code> is run
+ * again with <code>block</code> equal to <code>A.rows() / 2</code>. Once the
  * Cholesky Decomposition is computed, the full matrix cholesky is created
  * by propogating the cholesky forward as given in the reference report below.
  *
  * For a full guide to how this works
  * see the Cholesy decompostion chapter in the  reference report
  * <a href="https://goo.gl/6kWkJ5"> here</a>.
  * @param A Symmetric matrix on the GPU.
- * @param block Size of the block used to compute the cholesky decomposition.
- * @param divisor Proportion to divide the submatrix by at each recursive step.
- * @param min_block The amount that block is checked against to decide
- *  whether to continue the recursion or to perform the cholesky.
+ * @param min_block The minimum block size to execute the cholesky on.
  * @return Square root of matrix on the GPU.
  * @throw std::domain_error if m is not
  *  positive definite (if m has more than 0 elements)
  */
-inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block,
-                                     const int divisor, const int min_block) {
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int min_block) {
   auto offset = 0;
+  auto block = floor(A.rows() / 2);
   // NOTE: The code in this section follows the naming conventions
   // in the report linked in the docs.
   matrix_gpu A_11(block, block);
   // Repeats the blocked cholesky decomposition until the size of the remaining
   // submatrix is smaller or equal to the block size
-  while ((offset + block) < (A.rows())) {
+  if ((offset + block) < (A.rows())) {
```

Comment:
Not to be difficult here but I keep trying to figure out why this is looking weird to me and I think it's because the `_recursion` version should actually be the `cholesky_decompose()` entry point that divides up the work and the version that does the actual GPU stuff should have the `_whatever` tag. That way you can get rid of this `if` and the block below about `remaining_rows` where you do essentially the same thing as in the block above. I think. Here is sort of what I'm imagining in Python:
![image](https://user-images.githubusercontent.com/805356/51724459-7f200c00-202b-11e9-93f9-5838e5a0376a.png)

Reply:
Okay so with the above I think I was able to make what you are looking for. Looks a lot nicer now!

I'm still kicking the tires on it to make sure it makes sense but we got rid of the if statement blocks and the while loop so now it's has a pretty generic recurrent structure.